### ISSUE TITLE
Docs: improve general methods documentation

### ISSUE THREAD

**Issue by Valentin271** [ratatui-org/ratatui#386]
## Problem

There is not enough documentation in the widgets' methods (and other relative elements such as `Title` or `Padding`). While most of them are simple and require almost no documentation, some other can be ambiguous.

This was a major pain point for me while discovering the library (as a user and as a contributor).

## Solution

Add documentation

## Alternatives
_None_

## Additional context

I'm totally down to work on it, I just want to advertise this work for users and possibly interested contributors.

Elements missing documentation, grouped by widget :

- [x] `BarChart`, `Bar` (#449)
- [x] `Block`, `Padding`, `BorderType` (#469)
- [ ] `Chart`, `Axis`, `Dataset` (#696 )
- [x] `Gauge` (#514)
- [x] `List`, `ListState`, `ListItem` (#669)
- [x] `Paragraph` (#467)
- [x] `Sparkline`, `RenderDirection` (#648)
- [x] `Table` (#471, #638)
- [x] `Tabs` (#535)

There are other elements that need documentation, but I think they can be addressed later in an other issue.

---

Would it be useful to warn (and deny, once completed) on missing doc? See https://doc.rust-lang.org/rustdoc/lints.html#missing_docs

**Comment by Valentin271** [comment]
Also, I think _documentation_ and maybe _good first issue_ are good labels here. Thanks

**Comment by joshka** [comment]
>Would it be useful to warn (and deny, once completed) on missing doc? See https://doc.rust-lang.org/rustdoc/lints.html#missing_docs

Yes. Definitely! (Probably worth doing per file once they are documented well)

**Comment by joshka** [comment]
P.S. I recommend using an LLM (e.g. OpenAI / Llama / Claude) to help make this quicker to implement. Obviously validate the output and don't just accept AI hallucinations as gospel :)

Perhaps add some general doc guidelines to this issue or point to some existing ones if you're aware of them.
Some key points I'd worry about are (please add some more if you have them):
- keep line length to 100 chars (to make it easy to maintain) - the [vscode rewrap extension](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap) is good for this
- first line is summary, second line blank, third onwards is more detail.
- the main type doc comment should talk about the general features that the widget supports and introduce the concepts pointing to the various methods. Focus on interaction with various features and giving enough information that helps understand why you might want something. Also point out alternates
- any example code is as simple as possible

Also, another way to contribute to the documentation is to improve the examples so they cover each feature a little better. the examples are scraped automatically to provide examples in the docs page (use `cargo +nightly doc -Zunstable-options -Zrustdoc-scrape-examples --all-features --open` or run `bacon` and press `d`). A recently updated example is the [block example](https://github.com/ratatui-org/ratatui/tree/main/examples#block-blockrs), which gives the following output without any changes to the block's comments:

<img width="1001" alt="image" src="https://github.com/ratatui-org/ratatui/assets/381361/70ef7ef9-a04f-4ef0-a1d5-310ae7707299">

I've been thinking a bit about how and whether to add images to the doc comments. Like perhaps it would be worth chopping the block example up and putting an image of each part of it directly into the doc comment (this is fairly simple to automate with VHS, but requires a lot of effort to do at scale).

**Comment by DreadedHippy** [comment]
Hi,

I'm interested in fixing this issue. I'd like to start with "Paragraph" but I'm not very sure what needs changing. Could you provide more detail please?

It is unchecked in the list above but it doesn't seem to be missing documentation

**Comment by Valentin271** [comment]
Hello, it seems the `Paragraph` has been documented a little bit more since I wrote the list. Still it is missing one docstring for the [`alignment` method](https://github.com/ratatui-org/ratatui/blob/343c6cdc47c4fe38e64633d982aa413be356fb90/src/widgets/paragraph.rs#L199). (You can add `#![warn(missing_docs)]` at the top of the file to see it).

The main type doc can also be improved by referencing to the different concepts of `Paragraph`.

You can also look [here](https://github.com/ratatui-org/ratatui/blob/main/CONTRIBUTING.md#documentation) for some guidelines.

**Comment by DreadedHippy** [comment]
Alright, thank you very much for this.

I'll begin documenting Paragraph and move from there

**Comment by joshka** [comment]
Some guidelines to consider:
- https://rust-lang.github.io/api-guidelines/documentation.html
- https://github.com/jntrnr/rfcs/blob/front_page_styleguide/text/0000-api-doc-frontpage-styleguide.md
- https://github.com/rust-lang/rfcs/blob/master/text/1574-more-api-documentation-conventions.md
- https://github.com/ratatui-org/ratatui/blob/main/CONTRIBUTING.md#documentation

How about writing a concise checklist based on the above links (customized for Ratatui) that helps evaluate whether the docs are complete?

**Comment by DreadedHippy** [comment]
Alright, I will do this

**Comment by joshka** [comment]
Also no need to close https://github.com/ratatui-org/ratatui/pull/467 - this is fine to add as is with the suggested changes.

**Comment by DreadedHippy** [comment]
> Some guidelines to consider:
> 
> * https://rust-lang.github.io/api-guidelines/documentation.html
> * https://github.com/jntrnr/rfcs/blob/front_page_styleguide/text/0000-api-doc-frontpage-styleguide.md
> * https://github.com/rust-lang/rfcs/blob/master/text/1574-more-api-documentation-conventions.md
> * https://github.com/ratatui-org/ratatui/blob/main/CONTRIBUTING.md#documentation
> 
> How about writing a concise checklist based on the above links (customized for Ratatui) that helps evaluate whether the docs are complete?

[ ] Are all methods documented?
[ ] Are all enums documented?
[ ] Are all documented types hyperlinked?
[ ] Is the example given the minimum viable example?

Do you mean something like the above?
@joshka

**Comment by joshka** [comment]
> Do you mean something like the above? @joshka
Yep

- [ ] All types, traits, methods, and enums are documented (add `#![warn(missing_docs)]` once a module is complete)
- [ ] All mentions of a type or function are hyperlinked like ``[`Layout`]``
  To make it easier to read the docs in the doc comment, use references at the end of the comment: ``[`Layout`]: crate::layout::Layout``
- [ ] Interactions between methods are documented at the type level
- [ ] Methods include Panic / Error details
- [ ] Every public module, trait, type, and function has a simple example. The intent of the example is to show *why someone would want to use* the item, not just how to use it.
- [ ] Examples follow conventions:
  - use a `` ```plain `` code fence to show rendered widgets (unless obvious)
  - hide imports in method examples (except: types not part of the prelude / not obvious from context)
  - use `use ratatui::prelude::*` to avoid common imports
  - use `Style::new().red()` rather than `Style::default().fg(Color::Red)`
  - use `?` rather than `unwrap`

**Comment by TieWay59** [comment]
I have a simple off-topic question: is there a way to `wrap` the document-comment automatically? I found it annoying if I wanted to edit sentences of the doc but it led to a heavy job of manually reformating the comments.

**Comment by Valentin271** [comment]
nvim with rust-analyzer does it automatically for me. For vscode there is a [rewrap](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap) extension. There are probably similar extension for other editors.

**Comment by farmeroy** [comment]
Is the list of widgets that still need documentation up to date? I was working with various List widgets and Table and would be happy to try to improve those docs

**Comment by joshka** [comment]
> Is the list of widgets that still need documentation up to date? I was working with various List widgets and Table and would be happy to try to improve those docs

I wouldn't count on it being so - the latest alpha release docs are available at https://docs.rs/ratatui/0.23.1-alpha.6/ratatui/index.html - feel fee to jump in wherever things are missing or could be better explained.

**Comment by Valentin271** [comment]
I've just checked and it is indeed up to date, those widgets are still missing documentation.
